api/src/main/java/io/druid/data/input/impl/PrefetchableTextFilesFirehoseFactory.java

```diff
@@ -0,0 +1,375 @@
+
+          {
+            fetchIfNeeded(remainingCachedBytes);
+          }
+
+          @Override
+          public boolean hasNext()
+          {
+            return cacheFileIterator.hasNext()
+                   || !fetchFiles.isEmpty()
+                   || nextFetchIndex < objects.size();
+          }
+
+          private void fetchIfNeeded(long remainingBytes)
+          {
+            if ((fetchFuture == null || fetchFuture.isDone())
+                && remainingBytes <= prefetchTriggerBytes) {
+              fetchFuture = fetchExecutor.submit(
+                  () -> {
+                    fetch(objects, temporaryDirectory);
+                    return null;
+                  }
+              );
+            }
+          }
+
+          @Override
+          public LineIterator next()
+          {
+            if (!hasNext()) {
+              throw new NoSuchElementException();
+            }
+
+            // If fetch() fails, hasNext() always returns true because nextFetchIndex must be smaller than the number
+            // of objects which means next() is always called. The below block checks that fetch() threw an exception
+            // and propagates it if exists.
+            if (fetchFuture != null && fetchFuture.isDone()) {
+              try {
+                fetchFuture.get();
+              }
+              catch (InterruptedException | ExecutionException e) {
+                Throwables.propagate(e);
+              }
+            }
+
+            final File fetchedFile;
+            final Closeable closeable;
+            // Check cache first
+            if (cacheFileIterator.hasNext()) {
+              fetchedFile = cacheFileIterator.next();
+              remainingCachedBytes -= fetchedFile.length();
+              fetchIfNeeded(remainingCachedBytes);
+              closeable = () -> {
+              };
+            } else {
+              if (!fetchFiles.isEmpty()) {
+                // If there are already fetched files, use them
+                fetchedFile = fetchFiles.poll();
+                fetchIfNeeded(fetchedBytes.get());
+              } else {
+                // Otherwise, wait for fetching
+                try {
+                  fetchIfNeeded(fetchedBytes.get());
+                  fetchedFile = fetchFiles.poll(fetchTimeout, TimeUnit.MILLISECONDS);
+                  if (fetchedFile == null) {
+                    throw new RuntimeException(new TimeoutException());
+                  }
+                }
+                catch (InterruptedException e) {
+                  throw Throwables.propagate(e);
+                }
+              }
+              closeable = () -> {
+                final long fileSize = fetchedFile.length();
+                fetchedFile.delete();
+                fetchedBytes.addAndGet(-fileSize);
+              };
+            }
+
+            try {
+              InputStream stream = FileUtils.openInputStream(fetchedFile);
```

Comment:
```AbstractTextFilesFirehoseFactory.openStream()``` is used for fetching data from remote storage in ```PrefetchableTextFilesFirehoseFactory.download()```. Once a file is fetched, it is opened with ```FileUtils.openInputStream()```.